{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a665627e",
   "metadata": {},
   "source": [
    "# Sequence to function\n",
    "\n",
    "## The problem\n",
    "The problem we are working on is called sequence to function. Ideally this means we have a sequence and we infer some kind of function\n",
    "\n",
    "### The sequence: \n",
    "To me this means a few different things. \n",
    "- Gene sequence mutations\n",
    "- Gene and Protein Orthologs\n",
    "- Post-translational modification\n",
    "\n",
    "### The function: \n",
    "This also could mean a few things. My mind initially went to physiology, but now realized function could be alot broader. For example simple ligand enzyme binding could be related to funciton. The pathway that the protein is involved in, the regulation\n",
    "- A change in ligand binding interactions (Initial GPCR activation)\n",
    "- A change in metabolitic secondary activity (GPCRs downstream)\n",
    "\n",
    "## The solution\n",
    "When thinking about a solution to the hackathon problem. Ryan has suggested, could we start with building a full profile with only two data categories (for the sequence). To me this means: \n",
    "- Gene name, mutation/ortholog\n",
    "- Protein name, post-translational modification\n",
    "\n",
    "Given those could we pull out somekind of knowledge graph that will allow us to relate it to currently known research. This is where we will need the __agent__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80e1f8c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.fetch_data import split_colon_list, fetch_uniprot_data\n",
    "from scripts.epmc_utils import fetch_epmc_batch, fetch_epmc_batch_save_json, save_dataframe_rows_as_json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "322e207b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists\n"
     ]
    }
   ],
   "source": [
    "#This simple script creates the JSON directory if it doesn't exist. We are adding this to .gitignore, maybe it is too heavy?\n",
    "if not os.path.isdir(\"../data/corpus/\"):\n",
    "    os.makedirs(\"../data/corpus/\", exist_ok = True)\n",
    "    print(\"Making a new directory\")\n",
    "else:\n",
    "    print(\"Directory already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6469f8c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching UniProt data for CCR1\n",
      "Fetching UniProt data for CCR2\n",
      "Fetching UniProt data for CCR5\n",
      "Fetching UniProt data for CCR7\n",
      "Fetching UniProt data for APOE\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "161"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Curate the uniprot data\n",
    "uniprot_data = fetch_uniprot_data([\"CCR1\", \"CCR2\", \"CCR5\", \"CCR7\", \"APOE\"])\n",
    "#uniprot_data.to_csv(\"../data/processed/uniprot_specific_output.csv\", index=False)\n",
    "\n",
    "# #OR just open the output\n",
    "# uniprot_data =pd.read_csv(\"../data/processed/uniprot_output.csv\")\n",
    "# display(uniprot_data.head())\n",
    "\n",
    "citation_list = [\n",
    "    item\n",
    "    for citation_title in uniprot_data.citation_titles\n",
    "    for item in split_colon_list(citation_title)\n",
    "]\n",
    "len(citation_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c8e56b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching EPMC metadata for item 1 of 161: Molecular cloning, functional expression, and signaling characteristics of a C-C chemokine receptor.\n",
      "Fetching EPMC metadata for item 2 of 161: Structure and functional expression of the human macrophage inflammatory protein 1 alpha/RANTES receptor.\n",
      "Fetching EPMC metadata for item 3 of 161: Molecular cloning of cDNAs encoding a LD78 receptor and putative leukocyte chemotactic peptide receptors.\n",
      "Fetching EPMC metadata for item 4 of 161: The status, quality, and expansion of the NIH full-length cDNA project: the Mammalian Gene Collection (MGC).\n",
      "Fetching EPMC metadata for item 5 of 161: Human LZIP binds to CCR1 and differentially affects the chemotactic activities of CCR1-dependent chemokines.\n",
      "Fetching EPMC metadata for item 6 of 161: Differential chemokine activation of CC chemokine receptor 1-regulated pathways: ligand selective activation of Galpha 14-coupled pathways.\n",
      "Fetching EPMC metadata for item 7 of 161: Proteolytic activation of alternative CCR1 ligands in inflammation.\n",
      "Fetching EPMC metadata for item 8 of 161: Human LZIP induces monocyte CC chemokine receptor 2 expression leading to enhancement of monocyte chemoattractant protein 1/CCL2-induced cell migration.\n",
      "Fetching EPMC metadata for item 9 of 161: CXCL4/Platelet Factor 4 is an agonist of CCR1 and drives human monocyte migration.\n",
      "Fetching EPMC metadata for item 10 of 161: Identification and mechanism of G protein-biased ligands for chemokine receptor CCR1.\n",
      "Fetching EPMC metadata for item 11 of 161: Molecular cloning and functional expression of two monocyte chemoattractant protein 1 receptors reveals alternative splicing of the carboxyl-terminal tails.\n",
      "Fetching EPMC metadata for item 12 of 161: cDNA cloning and functional expression of a human monocyte chemoattractant protein 1 receptor.\n",
      "Fetching EPMC metadata for item 13 of 161: Organization and differential expression of the human monocyte chemoattractant protein 1 receptor gene. Evidence for the role of the carboxyl-terminal tail in receptor trafficking.\n",
      "Fetching EPMC metadata for item 14 of 161: Complete sequencing and characterization of 21,243 full-length human cDNAs.\n",
      "Fetching EPMC metadata for item 15 of 161: The DNA sequence, annotation and analysis of human chromosome 3.\n",
      "Fetching EPMC metadata for item 16 of 161: The status, quality, and expansion of the NIH full-length cDNA project: the Mammalian Gene Collection (MGC).\n",
      "Fetching EPMC metadata for item 17 of 161: IL-2-Regulated Expression of the Monocyte Chemotactic Protein-1 Receptor (CCR2) in Human NK Cells.\n",
      "Skipping existing file: IL-2-Regulated Expression of the Monocyte Chemotactic Protein-1 Receptor (CCR2) in Human NK Cells. (filename: EPMC_item0016.json)\n",
      "Fetching EPMC metadata for item 18 of 161: The chemokine monocyte chemotactic protein 1 triggers Janus kinase 2 activation and tyrosine phosphorylation of the CCR2B receptor.\n",
      "Fetching EPMC metadata for item 19 of 161: Monocyte chemoattractant protein-1-induced CCR2B receptor desensitization mediated by the G protein-coupled receptor kinase 2.\n",
      "Fetching EPMC metadata for item 20 of 161: HIV-1 Tat protein mimicry of chemokines.\n",
      "Fetching EPMC metadata for item 21 of 161: Monocyte chemotactic protein-1 receptor CCR2B is a glycoprotein that has tyrosine sulfation in a conserved extracellular N-terminal region.\n",
      "Fetching EPMC metadata for item 22 of 161: Pivotal function for cytoplasmic protein FROUNT in CCR2-mediated monocyte chemotaxis.\n",
      "Fetching EPMC metadata for item 23 of 161: Human LZIP induces monocyte CC chemokine receptor 2 expression leading to enhancement of monocyte chemoattractant protein 1/CCL2-induced cell migration.\n",
      "Fetching EPMC metadata for item 24 of 161: Tyrosine sulfation of chemokine receptor CCR2 enhances interactions with both monomeric and dimeric forms of the chemokine monocyte chemoattractant protein-1 (MCP-1).\n",
      "Fetching EPMC metadata for item 25 of 161: Structural basis for the interaction of human beta-defensin 6 and its putative chemokine receptor CCR2 and breast cancer microvesicles.\n",
      "Fetching EPMC metadata for item 26 of 161: Structural basis for the binding of the membrane-proximal C-terminal region of chemokine receptor CCR2 with the cytosolic regulator FROUNT.\n",
      "Fetching EPMC metadata for item 27 of 161: Structure of CC chemokine receptor 2 with orthosteric and allosteric antagonists.\n",
      "Fetching EPMC metadata for item 28 of 161: Contrasting genetic influence of CCR2 and CCR5 variants on HIV-1 infection and disease progression.\n",
      "Fetching EPMC metadata for item 29 of 161: Genealogy of the CCR5 locus and chemokine system gene variants associated with altered rates of HIV-1 disease progression.\n",
      "Fetching EPMC metadata for item 30 of 161: Human inherited CCR2 deficiency underlies progressive polycystic lung disease.\n",
      "Fetching EPMC metadata for item 31 of 161: Molecular cloning and functional expression of a new human CC-chemokine receptor gene.\n",
      "Fetching EPMC metadata for item 32 of 161: Molecular cloning and functional characterization of a novel human CC chemokine receptor (CCR5) for RANTES, MIP-1beta, and MIP-1alpha.\n",
      "Fetching EPMC metadata for item 33 of 161: Cloning and functional expression of CC CKR5, a human monocyte CC chemokine receptor selective for MIP-1(alpha), MIP-1(beta), and RANTES.\n",
      "Fetching EPMC metadata for item 34 of 161: Polymorphisms in the CCR5 genes of African green monkeys and mice implicate specific amino acids in infections by simian and human immunodeficiency viruses.\n",
      "Fetching EPMC metadata for item 35 of 161: HIV type 1 subtypes, coreceptor usage, and CCR5 polymorphism.\n",
      "Fetching EPMC metadata for item 36 of 161: The human CC chemokine receptor 5 (CCR5) gene. Multiple transcripts with 5'-end heterogeneity, dual promoter usage, and evidence for polymorphisms within the regulatory regions and noncoding exons.\n",
      "Fetching EPMC metadata for item 37 of 161: Novel variant of the CCR5 gene in a Vietnamese population.\n",
      "Fetching EPMC metadata for item 38 of 161: cDNA clones of human proteins involved in signal transduction sequenced by the Guthrie cDNA resource center (www.cdna.org).\n",
      "Skipping existing file: cDNA clones of human proteins involved in signal transduction sequenced by the Guthrie cDNA resource center (www.cdna.org). (filename: EPMC_item0037.json)\n",
      "Fetching EPMC metadata for item 39 of 161: The DNA sequence, annotation and analysis of human chromosome 3.\n",
      "Fetching EPMC metadata for item 40 of 161: The status, quality, and expansion of the NIH full-length cDNA project: the Mammalian Gene Collection (MGC).\n",
      "Fetching EPMC metadata for item 41 of 161: Homozygous defect in HIV-1 coreceptor accounts for resistance of some multiply-exposed individuals to HIV-1 infection.\n",
      "Fetching EPMC metadata for item 42 of 161: Identification of a major co-receptor for primary isolates of HIV-1.\n",
      "Fetching EPMC metadata for item 43 of 161: HIV-1 entry into CD4+ cells is mediated by the chemokine receptor CC-CKR-5.\n",
      "Fetching EPMC metadata for item 44 of 161: Resistance to HIV-1 infection in caucasian individuals bearing mutant alleles of the CCR-5 chemokine receptor gene.\n",
      "Fetching EPMC metadata for item 45 of 161: A conserved HIV gp120 glycoprotein structure involved in chemokine receptor binding.\n",
      "Fetching EPMC metadata for item 46 of 161: Tyrosine sulfation of the amino terminus of CCR5 facilitates HIV-1 entry.\n",
      "Fetching EPMC metadata for item 47 of 161: Differential effects of CC chemokines on CC chemokine receptor 5 (CCR5) phosphorylation and identification of phosphorylation sites on the CCR5 carboxyl terminus.\n",
      "Fetching EPMC metadata for item 48 of 161: Extracellular cysteines of CCR5 are required for chemokine binding, but dispensable for HIV-1 coreceptor activity.\n",
      "Fetching EPMC metadata for item 49 of 161: Palmitoylation of CCR5 is critical for receptor trafficking and efficient activation of intracellular signaling pathways.\n",
      "Fetching EPMC metadata for item 50 of 161: Characterization of sequence determinants within the carboxyl-terminal domain of chemokine receptor CCR5 that regulate signaling and receptor internalization.\n",
      "Fetching EPMC metadata for item 51 of 161: Sialylated O-glycans and sulfated tyrosines in the NH2-terminal domain of CC chemokine receptor 5 contribute to high affinity binding of chemokines.\n",
      "Fetching EPMC metadata for item 52 of 161: JM4 is a four-transmembrane protein binding to the CCR5 receptor.\n",
      "Fetching EPMC metadata for item 53 of 161: G protein-coupled receptor kinases promote phosphorylation and beta-arrestin-mediated internalization of CCR5 homo- and hetero-oligomers.\n",
      "Fetching EPMC metadata for item 54 of 161: CCR5 deficiency increases risk of symptomatic West Nile virus infection.\n",
      "Fetching EPMC metadata for item 55 of 161: Human cytomegalovirus-encoded UL33 and UL78 heteromerize with host CCR5 and CXCR4 impairing their HIV coreceptor activity.\n",
      "Fetching EPMC metadata for item 56 of 161: CNIH4 interacts with newly synthesized GPCR and controls their export from the endoplasmic reticulum.\n",
      "Fetching EPMC metadata for item 57 of 161: Tag7-Mts1 Complex Induces Lymphocytes Migration via CCR5 and CXCR3 Receptors.\n",
      "Fetching EPMC metadata for item 58 of 161: Structure modeling of the chemokine receptor CCR5: implications for ligand binding and selectivity.\n",
      "Fetching EPMC metadata for item 59 of 161: Structural and functional characterization of the human CCR5 receptor in complex with HIV gp120 envelope glycoprotein and CD4 receptor by molecular modeling studies.\n",
      "Fetching EPMC metadata for item 60 of 161: The conformation and orientation of a 27-residue CCR5 peptide in a ternary complex with HIV-1 gp120 and a CD4-mimic peptide.\n",
      "Fetching EPMC metadata for item 61 of 161: Structure of the CCR5 chemokine receptor-HIV entry inhibitor maraviroc complex.\n",
      "Fetching EPMC metadata for item 62 of 161: The extent of genetic variation in the CCR5 gene.\n",
      "Fetching EPMC metadata for item 63 of 161: Novel alleles of the chemokine-receptor gene CCR5.\n",
      "Fetching EPMC metadata for item 64 of 161: A homologous naturally occurring mutation in Duffy and CCR5 leading to reduced receptor expression.\n",
      "Fetching EPMC metadata for item 65 of 161: Persistent resistance to HIV-1 infection in CD4 T cells from exposed uninfected Vietnamese individuals is mediated by entry and post-entry blocks.\n",
      "Fetching EPMC metadata for item 66 of 161: Shared and distinct genetic variants in type 1 diabetes and celiac disease.\n",
      "Fetching EPMC metadata for item 67 of 161: Epstein-Barr virus-induced genes: first lymphocyte-specific G protein-coupled peptide receptors.\n",
      "Fetching EPMC metadata for item 68 of 161: Cloning of human and mouse EBI1, a lymphoid-specific G-protein-coupled receptor encoded on human chromosome 17q12-q21.2.\n",
      "Fetching EPMC metadata for item 69 of 161: cDNA clones of human proteins involved in signal transduction sequenced by the Guthrie cDNA resource center (www.cdna.org).\n",
      "Skipping existing file: cDNA clones of human proteins involved in signal transduction sequenced by the Guthrie cDNA resource center (www.cdna.org). (filename: EPMC_item0068.json)\n",
      "Fetching EPMC metadata for item 70 of 161: The status, quality, and expansion of the NIH full-length cDNA project: the Mammalian Gene Collection (MGC).\n",
      "Fetching EPMC metadata for item 71 of 161: Synthesis, intracellular processing, and signal peptide of human apolipoprotein E.\n",
      "Fetching EPMC metadata for item 72 of 161: Human apolipoprotein E mRNA. cDNA cloning and nucleotide sequencing of a new variant.\n",
      "Fetching EPMC metadata for item 73 of 161: Nucleotide sequence and structure of the human apolipoprotein E gene.\n",
      "Fetching EPMC metadata for item 74 of 161: Genotyping and sequence analysis of apolipoprotein E isoforms.\n",
      "Fetching EPMC metadata for item 75 of 161: Sequencing of 42kb of the APO E-C2 gene cluster reveals a new gene: PEREC1.\n",
      "Fetching EPMC metadata for item 76 of 161: Sequence diversity and large-scale typing of SNPs in the human apolipoprotein E gene.\n",
      "Fetching EPMC metadata for item 77 of 161: Complete sequencing and characterization of 21,243 full-length human cDNAs.\n",
      "Fetching EPMC metadata for item 78 of 161: The status, quality, and expansion of the NIH full-length cDNA project: the Mammalian Gene Collection (MGC).\n",
      "Fetching EPMC metadata for item 79 of 161: A new apolipoprotein E variant (Gln46-->His).\n",
      "Skipping existing file: A new apolipoprotein E variant (Gln46-->His). (filename: EPMC_item0078.json)\n",
      "Fetching EPMC metadata for item 80 of 161: Identification and DNA sequence of a human apolipoprotein E cDNA clone.\n",
      "Fetching EPMC metadata for item 81 of 161: Human apolipoprotein E. The complete amino acid sequence.\n",
      "Fetching EPMC metadata for item 82 of 161: Cholesteryl ester and apolipoprotein E transfer between human high density lipoproteins and chylomicrons.\n",
      "Fetching EPMC metadata for item 83 of 161: Binding of a high reactive heparin to human apolipoprotein E: identification of two heparin-binding domains.\n",
      "Fetching EPMC metadata for item 84 of 161: Lipoproteins and their receptors in the central nervous system. Characterization of the lipoproteins in cerebrospinal fluid and identification of apolipoprotein B,E(LDL) receptors in the brain.\n",
      "Fetching EPMC metadata for item 85 of 161: Soluble LR11 competes with amyloid beta in binding to cerebrospinal fluid-high-density lipoprotein.\n",
      "Fetching EPMC metadata for item 86 of 161: Site-specific mutagenesis of human apolipoprotein E. Receptor binding activity of variants with single amino acid substitutions.\n",
      "Fetching EPMC metadata for item 87 of 161: Glycosylation of human apolipoprotein E. The carbohydrate attachment site is threonine 194.\n",
      "Fetching EPMC metadata for item 88 of 161: Low density lipoprotein receptor-related protein mediates uptake of cholesteryl esters derived from apoprotein E-enriched lipoproteins.\n",
      "Fetching EPMC metadata for item 89 of 161: Apolipoprotein E distribution among human plasma lipoproteins: role of the cysteine-arginine interchange at residue 112.\n",
      "Fetching EPMC metadata for item 90 of 161: Effects of exogenous apo E-3 and of cholesterol-enriched meals on the cellular metabolism of human chylomicrons and their remnants.\n",
      "Fetching EPMC metadata for item 91 of 161: Mechanisms of inhibition by apolipoprotein C of apolipoprotein E-dependent cellular metabolism of human triglyceride-rich lipoproteins through the low density lipoprotein receptor pathway.\n",
      "Fetching EPMC metadata for item 92 of 161: Discrete carboxyl-terminal segments of apolipoprotein E mediate lipoprotein association and protein oligomerization.\n",
      "Fetching EPMC metadata for item 93 of 161: Binding of human apolipoprotein E to synthetic amyloid beta peptide: isoform-specific effects and implications for late-onset Alzheimer disease.\n",
      "Fetching EPMC metadata for item 94 of 161: Isoform-specific interactions of apolipoprotein E with the microtubule-associated protein MAP2c: implications for Alzheimer's disease.\n",
      "Fetching EPMC metadata for item 95 of 161: Isoform-specific interactions of apolipoprotein E with microtubule-associated protein tau: implications for Alzheimer disease.\n",
      "Fetching EPMC metadata for item 96 of 161: Identification of glycoprotein 330 as an endocytic receptor for apolipoprotein J/clusterin.\n",
      "Fetching EPMC metadata for item 97 of 161: Apolipoprotein E-containing high density lipoprotein promotes neurite outgrowth and is a ligand for the low density lipoprotein receptor-related protein.\n",
      "Fetching EPMC metadata for item 98 of 161: Heparan sulfate proteoglycans participate in hepatic lipase and apolipoprotein E-mediated binding and uptake of plasma lipoproteins, including high density lipoproteins.\n",
      "Fetching EPMC metadata for item 99 of 161: The HepG2 extracellular matrix contains separate heparinase- and lipid-releasable pools of ApoE. Implications for hepatic lipoprotein metabolism.\n",
      "Fetching EPMC metadata for item 100 of 161: Specific regional transcription of apolipoprotein E in human brain neurons.\n",
      "Fetching EPMC metadata for item 101 of 161: Glycation of apolipoprotein E impairs its binding to heparin: identification of the major glycation site.\n",
      "Fetching EPMC metadata for item 102 of 161: Apolipoprotein E fragments present in Alzheimer's disease brains induce neurofibrillary tangle-like intracellular inclusions in neurons.\n",
      "Fetching EPMC metadata for item 103 of 161: Domains of apoE required for binding to apoE receptor 2 and to phospholipids: implications for the functions of apoE in the brain.\n",
      "Fetching EPMC metadata for item 104 of 161: Molecular interactions between apoE and ABCA1: impact on apoE lipidation.\n",
      "Fetching EPMC metadata for item 105 of 161: Enrichment of glycopeptides for glycan structure and attachment site identification.\n",
      "Fetching EPMC metadata for item 106 of 161: Decoding of lipoprotein-receptor interactions: properties of ligand binding modules governing interactions with apolipoprotein E.\n",
      "Fetching EPMC metadata for item 107 of 161: Glycosylation and sialylation of macrophage-derived human apolipoprotein E analyzed by SDS-PAGE and mass spectrometry: evidence for a novel site of glycosylation on Ser290.\n",
      "Fetching EPMC metadata for item 108 of 161: Initial characterization of the human central proteome.\n",
      "Fetching EPMC metadata for item 109 of 161: Resveratrol-induced changes of the human adipocyte secretion profile.\n",
      "Fetching EPMC metadata for item 110 of 161: Apolipoproteins E and AV mediate lipoprotein clearance by hepatic proteoglycans.\n",
      "Fetching EPMC metadata for item 111 of 161: LC-MS/MS characterization of O-glycosylation sites and glycan structures of human cerebrospinal fluid glycoproteins.\n",
      "Fetching EPMC metadata for item 112 of 161: ApoE influences amyloid-beta (Abeta) clearance despite minimal apoE/Abeta association in physiological conditions.\n",
      "Fetching EPMC metadata for item 113 of 161: An enzyme assisted RP-RPLC approach for in-depth analysis of human liver phosphoproteome.\n",
      "Fetching EPMC metadata for item 114 of 161: Apolipoprotein E likely contributes to a maturation step of infectious hepatitis C virus particles and interacts with viral envelope glycoproteins.\n",
      "Fetching EPMC metadata for item 115 of 161: A single kinase generates the majority of the secreted phosphoproteome.\n",
      "Fetching EPMC metadata for item 116 of 161: Apolipoprotein E Regulates Amyloid Formation within Endosomes of Pigment Cells.\n",
      "Fetching EPMC metadata for item 117 of 161: ApoE2, ApoE3, and ApoE4 Differentially Stimulate APP Transcription and Abeta Secretion.\n",
      "Fetching EPMC metadata for item 118 of 161: Regulation of Apolipoprotein E Trafficking by Hepatitis C Virus-Induced Autophagy.\n",
      "Fetching EPMC metadata for item 119 of 161: LILRB4 signalling in leukaemia cells mediates T cell suppression and tumour infiltration.\n",
      "Fetching EPMC metadata for item 120 of 161: Three-dimensional structure of the LDL receptor-binding domain of human apolipoprotein E.\n",
      "Fetching EPMC metadata for item 121 of 161: Human apolipoprotein E. Role of arginine 61 in mediating the lipoprotein preferences of the E3 and E4 isoforms.\n",
      "Fetching EPMC metadata for item 122 of 161: Salt bridge relay triggers defective LDL receptor binding by a mutant apolipoprotein.\n",
      "Fetching EPMC metadata for item 123 of 161: Novel mechanism for defective receptor binding of apolipoprotein E2 in type III hyperlipoproteinemia.\n",
      "Fetching EPMC metadata for item 124 of 161: Conformational flexibility in the apolipoprotein E amino-terminal domain structure determined from three new crystal forms: implications for lipid binding.\n",
      "Fetching EPMC metadata for item 125 of 161: Interaction of the N-terminal domain of apolipoprotein E4 with heparin.\n",
      "Fetching EPMC metadata for item 126 of 161: Structure of the minimal interface between ApoE and LRP.\n",
      "Fetching EPMC metadata for item 127 of 161: Identification of human apolipoprotein E variant gene: apolipoprotein E7 (Glu244,245----Lys244,245).\n",
      "Fetching EPMC metadata for item 128 of 161: Molecular cloning of a human apolipoprotein E variant: E5 (Glu-3-->Lys).\n",
      "Fetching EPMC metadata for item 129 of 161: Apolipoprotein E3-Leiden contains a seven-amino acid insertion that is a tandem repeat of residues 121-127.\n",
      "Fetching EPMC metadata for item 130 of 161: Identification and characterization of a new variant of apolipoprotein E (apo E-Kochi).\n",
      "Fetching EPMC metadata for item 131 of 161: Apolipoprotein E2-Dunedin (228 Arg replaced by Cys): an apolipoprotein E2 variant with normal receptor-binding activity.\n",
      "Fetching EPMC metadata for item 132 of 161: Apolipoprotein E-4 Philadelphia (Glu-13-->Lys,Arg-145-->Cys). Homozygosity for two rare point mutations in the apolipoprotein E gene combined with severe type III hyperlipoproteinemia.\n",
      "Fetching EPMC metadata for item 133 of 161: Characterization of a new apolipoprotein E5 variant detected in two French-Canadian subjects.\n",
      "Fetching EPMC metadata for item 134 of 161: Site-directed mutagenesis of an apolipoprotein E mutant, apo E5(Glu3----Lys) and its binding to low density lipoprotein receptors.\n",
      "Fetching EPMC metadata for item 135 of 161: Familial apolipoprotein E deficiency and type III hyperlipoproteinemia due to a premature stop codon in the apolipoprotein E gene.\n",
      "Fetching EPMC metadata for item 136 of 161: Characterization of five new mutants in the carboxyl-terminal domain of human apolipoprotein E: no cosegregation with severe hyperlipidemia.\n",
      "Fetching EPMC metadata for item 137 of 161: Characterization of the gene for apolipoprotein E5-Frankfurt (Gln81->Lys, Cys112->Arg) by polymerase chain reaction, restriction isotyping, and temperature gradient gel electrophoresis.\n",
      "Fetching EPMC metadata for item 138 of 161: Preferential association of apolipoprotein E Leiden with very low density lipoproteins of human plasma.\n",
      "Fetching EPMC metadata for item 139 of 161: Gene dose of apolipoprotein E type 4 allele and the risk of Alzheimer's disease in late onset families.\n",
      "Fetching EPMC metadata for item 140 of 161: Common and rare genotypes of human apolipoprotein E determined by specific restriction profiles of polymerase chain reaction-amplified DNA.\n",
      "Fetching EPMC metadata for item 141 of 161: Dominant expression of type III hyperlipoproteinemia. Pathophysiological insights derived from the structural and kinetic characteristics of ApoE-1 (Lys146-->Glu).\n",
      "Fetching EPMC metadata for item 142 of 161: Characterization of a novel variant of apolipoprotein E, E2 Fukuoka (Arg-224 --> Gln) in a hyperlipidemic patient with xanthomatosis.\n",
      "Fetching EPMC metadata for item 143 of 161: Apolipoprotein E Sendai (arginine 145-->proline): a new variant associated with lipoprotein glomerulopathy.\n",
      "Fetching EPMC metadata for item 144 of 161: Apolipoprotein E R112\n",
      "Fetching EPMC metadata for item 145 of 161: R251G: a carboxy-terminal variant found in patients with hyperlipidemia and coronary heart disease.\n",
      "Fetching EPMC metadata for item 146 of 161: A novel apolipoprotein E mutation, E2 (Arg25Cys), in lipoprotein glomerulopathy.\n",
      "Fetching EPMC metadata for item 147 of 161: Virus-mediated transduction of apolipoprotein E (ApoE)-sendai develops lipoprotein glomerulopathy in ApoE-deficient mice.\n",
      "Fetching EPMC metadata for item 148 of 161: Familial splenomegaly: macrophage hypercatabolism of lipoproteins associated with apolipoprotein E mutation [apolipoprotein E (delta149 Leu)].\n",
      "Fetching EPMC metadata for item 149 of 161: Apolipoprotein E3Basel: new insights into a highly conserved protein region.\n",
      "Fetching EPMC metadata for item 150 of 161: Association of extreme blood lipid profile phenotypic variation with 11 reverse cholesterol transport genes and 10 non-genetic cardiovascular disease risk factors.\n",
      "Fetching EPMC metadata for item 151 of 161: Variable expressivity of the clinical and biochemical phenotype associated with the apolipoprotein E p.Leu149del mutation.\n",
      "Fetching EPMC metadata for item 152 of 161: APOE Kyoto mutation in European Americans with lipoprotein glomerulopathy.\n",
      "Fetching EPMC metadata for item 153 of 161: Quantitative detection of single amino acid polymorphisms by targeted proteomics.\n",
      "Fetching EPMC metadata for item 154 of 161: Apolipoprotein E gene mutations in subjects with mixed hyperlipidemia and a clinical diagnosis of familial combined hyperlipidemia.\n",
      "Fetching EPMC metadata for item 155 of 161: APOE p.Leu167del mutation in familial hypercholesterolemia.\n",
      "Fetching EPMC metadata for item 156 of 161: Description of a large family with autosomal dominant hypercholesterolemia associated with the APOE p.Leu167del mutation.\n",
      "Fetching EPMC metadata for item 157 of 161: Global molecular analysis and APOE mutations in a cohort of autosomal dominant hypercholesterolemia patients in France.\n",
      "Fetching EPMC metadata for item 158 of 161: Genetic heterogeneity of apolipoprotein E and its influence on plasma lipid and lipoprotein levels.\n",
      "Fetching EPMC metadata for item 159 of 161: Apolipoprotein E: structure and function in lipid metabolism, neurobiology, and Alzheimer's diseases.\n",
      "Fetching EPMC metadata for item 160 of 161: Cell-specific production, secretion, and function of apolipoprotein E.\n",
      "Fetching EPMC metadata for item 161 of 161: ApoE-Isoform-Dependent SARS-CoV-2 Neurotropism and Cellular Response.\n",
      "Saved 157 UniProt citation JSON files.\n"
     ]
    }
   ],
   "source": [
    "uniprot_citations_df = fetch_epmc_batch_save_json(\n",
    "    items = citation_list,\n",
    "    directory = \"../data/corpus/\",\n",
    "    id_column=\"PMID\",\n",
    "    filename_prefix=\"EPMC_\",\n",
    "    indent=4,\n",
    "    drop_missing=True,\n",
    ")\n",
    "print(f\"Saved {len(uniprot_citations_df)} UniProt citation JSON files.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
