{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a665627e",
   "metadata": {},
   "source": [
    "# Sequence to function\n",
    "\n",
    "## The problem\n",
    "The problem we are working on is called sequence to function. Ideally this means we have a sequence and we infer some kind of function\n",
    "\n",
    "### The sequence: \n",
    "To me this means a few different things. \n",
    "- Gene sequence mutations\n",
    "- Gene and Protein Orthologs\n",
    "- Post-translational modification\n",
    "\n",
    "### The function: \n",
    "This also could mean a few things. My mind initially went to physiology, but now realized function could be alot broader. For example simple ligand enzyme binding could be related to funciton. The pathway that the protein is involved in, the regulation\n",
    "- A change in ligand binding interactions (Initial GPCR activation)\n",
    "- A change in metabolitic secondary activity (GPCRs downstream)\n",
    "\n",
    "## The solution\n",
    "When thinking about a solution to the hackathon problem. Ryan has suggested, could we start with building a full profile with only two data categories (for the sequence). To me this means: \n",
    "- Gene name, mutation/ortholog\n",
    "- Protein name, post-translational modification\n",
    "\n",
    "Given those could we pull out somekind of knowledge graph that will allow us to relate it to currently known research. This is where we will need the __agent__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "80e1f8c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.fetch_data import fetch_uniprot_data, split_colon_list\n",
    "from scripts.reference_scoring import (\n",
    "    collect_reference_network_for_genes,\n",
    "    score_reference_dataframe,\n",
    "    select_top_scoring_articles,\n",
    ")\n",
    "from scripts.epmc_utils import fetch_epmc, save_dataframe_rows_as_json\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "322e207b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists\n"
     ]
    }
   ],
   "source": [
    "#This simple script creates the JSON directory if it doesn't exist. We are adding this to .gitignore, maybe it is too heavy?\n",
    "if not os.path.isdir(\"../data/papers\"):\n",
    "    os.makedirs(\"../data/papers\", exist_ok = True)\n",
    "    print(\"Making a new directory\")\n",
    "else:\n",
    "    print(\"Directory already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6469f8c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pmids</th>\n",
       "      <th>dois</th>\n",
       "      <th>citation_titles</th>\n",
       "      <th>reviewed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCR2</td>\n",
       "      <td>P41597</td>\n",
       "      <td>C-C chemokine receptor type 2</td>\n",
       "      <td>MLSTSRSRFIRNTNESGEEVTTFFDYDYGAPCHKFDVKQIGAQLLP...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Molecular cloning and functional expression of...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol uniprot_id                   protein_name  \\\n",
       "0        CCR2     P41597  C-C chemokine receptor type 2   \n",
       "\n",
       "                                            sequence pmids  dois  \\\n",
       "0  MLSTSRSRFIRNTNESGEEVTTFFDYDYGAPCHKFDVKQIGAQLLP...  None  None   \n",
       "\n",
       "                                     citation_titles  reviewed  \n",
       "0  Molecular cloning and functional expression of...     False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Maybe here can go the parsing terms\n",
    "uniprot_data = fetch_uniprot_data([\"CCR2\"])\n",
    "display(uniprot_data)\n",
    "citation_list = split_colon_list(uniprot_data.citation_titles[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bced0ed7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMC</th>\n",
       "      <th>DOI</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>title</th>\n",
       "      <th>journal</th>\n",
       "      <th>year</th>\n",
       "      <th>source_url</th>\n",
       "      <th>source</th>\n",
       "      <th>abstract_text</th>\n",
       "      <th>full_text</th>\n",
       "      <th>full_text_abstract</th>\n",
       "      <th>function_signal</th>\n",
       "      <th>longevity_signal</th>\n",
       "      <th>year_score</th>\n",
       "      <th>functionality_score</th>\n",
       "      <th>longevity_score</th>\n",
       "      <th>composite_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>10.1038/ni1222</td>\n",
       "      <td>15995708</td>\n",
       "      <td>None</td>\n",
       "      <td>Pivotal function for cytoplasmic protein FROUN...</td>\n",
       "      <td>None</td>\n",
       "      <td>2005</td>\n",
       "      <td>https://europepmc.org/article/MED/15995708</td>\n",
       "      <td>MED</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Ligation of the chemokine receptor CCR2 on mon...</td>\n",
       "      <td>1.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.848284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.315947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>43448</td>\n",
       "      <td>10.1073/pnas.91.7.2752</td>\n",
       "      <td>8146186</td>\n",
       "      <td>PMC43448</td>\n",
       "      <td>Molecular cloning and functional expression of...</td>\n",
       "      <td>None</td>\n",
       "      <td>1994</td>\n",
       "      <td>https://europepmc.org/article/PMC/PMC43448</td>\n",
       "      <td>MED</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Monocyte chemoattractant protein 1 (MCP-1) is ...</td>\n",
       "      <td>1.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.848284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.309399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>10.1006/bbrc.1994.2049</td>\n",
       "      <td>8048929</td>\n",
       "      <td>None</td>\n",
       "      <td>cDNA cloning and functional expression of a hu...</td>\n",
       "      <td>None</td>\n",
       "      <td>1994</td>\n",
       "      <td>https://europepmc.org/article/MED/8048929</td>\n",
       "      <td>MED</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>A novel human G-protein-coupled seven-transmem...</td>\n",
       "      <td>1.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.848284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.309399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11198729</td>\n",
       "      <td>10.1016/j.cell.2024.05.021</td>\n",
       "      <td>38776920</td>\n",
       "      <td>PMC11198729</td>\n",
       "      <td>Human inherited CCR2 deficiency underlies prog...</td>\n",
       "      <td>None</td>\n",
       "      <td>2024</td>\n",
       "      <td>https://europepmc.org/article/PMC/PMC11198729</td>\n",
       "      <td>MED</td>\n",
       "      <td>None</td>\n",
       "      <td>pmc 0413066 2830 Cell Cell Cell 0092-8674 1097...</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5509255</td>\n",
       "      <td>10.1371/journal.pone.0181027</td>\n",
       "      <td>28704482</td>\n",
       "      <td>PMC5509255</td>\n",
       "      <td>Proteomic identification of proteins different...</td>\n",
       "      <td>None</td>\n",
       "      <td>2017</td>\n",
       "      <td>https://europepmc.org/article/PMC/PMC5509255</td>\n",
       "      <td>MED</td>\n",
       "      <td>None</td>\n",
       "      <td>PLoS One PLoS ONE plos plosone PLoS ONE 1932-6...</td>\n",
       "      <td>Reverse transcriptase activity of telomerase a...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.044444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        PMC                           DOI      PMID        PMCID  \\\n",
       "0      None                10.1038/ni1222  15995708         None   \n",
       "1     43448        10.1073/pnas.91.7.2752   8146186     PMC43448   \n",
       "2      None        10.1006/bbrc.1994.2049   8048929         None   \n",
       "3  11198729    10.1016/j.cell.2024.05.021  38776920  PMC11198729   \n",
       "4   5509255  10.1371/journal.pone.0181027  28704482   PMC5509255   \n",
       "\n",
       "                                               title journal  year  \\\n",
       "0  Pivotal function for cytoplasmic protein FROUN...    None  2005   \n",
       "1  Molecular cloning and functional expression of...    None  1994   \n",
       "2  cDNA cloning and functional expression of a hu...    None  1994   \n",
       "3  Human inherited CCR2 deficiency underlies prog...    None  2024   \n",
       "4  Proteomic identification of proteins different...    None  2017   \n",
       "\n",
       "                                      source_url source abstract_text  \\\n",
       "0     https://europepmc.org/article/MED/15995708    MED          None   \n",
       "1     https://europepmc.org/article/PMC/PMC43448    MED          None   \n",
       "2      https://europepmc.org/article/MED/8048929    MED          None   \n",
       "3  https://europepmc.org/article/PMC/PMC11198729    MED          None   \n",
       "4   https://europepmc.org/article/PMC/PMC5509255    MED          None   \n",
       "\n",
       "                                           full_text  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3  pmc 0413066 2830 Cell Cell Cell 0092-8674 1097...   \n",
       "4  PLoS One PLoS ONE plos plosone PLoS ONE 1932-6...   \n",
       "\n",
       "                                  full_text_abstract  function_signal  \\\n",
       "0  Ligation of the chemokine receptor CCR2 on mon...             1.25   \n",
       "1  Monocyte chemoattractant protein 1 (MCP-1) is ...             1.25   \n",
       "2  A novel human G-protein-coupled seven-transmem...             1.25   \n",
       "3                                               None             0.00   \n",
       "4  Reverse transcriptase activity of telomerase a...             0.00   \n",
       "\n",
       "   longevity_signal  year_score  functionality_score  longevity_score  \\\n",
       "0                 0    0.047619             0.848284              0.0   \n",
       "1                 0    0.031250             0.848284              0.0   \n",
       "2                 0    0.031250             0.848284              0.0   \n",
       "3                 0    0.500000             0.000000              0.0   \n",
       "4                 0    0.111111             0.000000              0.0   \n",
       "\n",
       "   composite_score  \n",
       "0         0.315947  \n",
       "1         0.309399  \n",
       "2         0.309399  \n",
       "3         0.200000  \n",
       "4         0.044444  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 20 UniProt citation JSON files.\n"
     ]
    }
   ],
   "source": [
    "uniprot_citation_records = [\n",
    "    fetch_epmc(citation)\n",
    "    for citation in citation_list\n",
    "]\n",
    "uniprot_citations_df = pd.DataFrame(uniprot_citation_records)\n",
    "\n",
    "uniprot_citations_df = score_reference_dataframe(\n",
    "    uniprot_citations_df,\n",
    "    delay=0.1,\n",
    "    include_fulltext=True,\n",
    ")\n",
    "display(uniprot_citations_df.head())\n",
    "\n",
    "saved_uniprot_citations = save_dataframe_rows_as_json(\n",
    "    uniprot_citations_df,\n",
    "    \"../data/papers/uniprot_citations\",\n",
    "    id_column=\"PMID\",\n",
    "    filename_prefix=\"uniprot_\",\n",
    "    indent=4,\n",
    "    drop_missing=True,\n",
    ")\n",
    "print(f\"Saved {len(saved_uniprot_citations)} UniProt citation JSON files.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "590052a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_network = collect_reference_network_for_genes(\n",
    "    uniprot_df=uniprot_data,\n",
    "    include=(\"references\", \"citations\"),\n",
    "    max_depth=1,\n",
    "    delay=0.1,\n",
    ")\n",
    "display(reference_network.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60956054",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_rows_scored = score_reference_dataframe(\n",
    "    reference_network,\n",
    "    delay=0.1,\n",
    "    include_fulltext=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f157d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_rows_scored[\"score\"] = reference_rows_scored[\"composite_score\"]\n",
    "saved_reference_files = save_dataframe_rows_as_json(\n",
    "    reference_rows_scored,\n",
    "    \"../data/papers/reference_list\",\n",
    "    id_column=\"PMID\",\n",
    "    filename_prefix=\"reference_\",\n",
    "    indent=4,\n",
    "    drop_missing=True,\n",
    ")\n",
    "print(f\"Saved {len(saved_reference_files)} reference JSON files.\")\n",
    "display(reference_rows_scored[[\"gene_symbol\", \"title\", \"year\", \"score\"]].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bd1bd64",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_references = 10\n",
    "\n",
    "if scored_network.empty:\n",
    "    top_references = scored_network\n",
    "    print(\"No articles available for ranking or export.\")\n",
    "else:\n",
    "    top_references = select_top_scoring_articles(\n",
    "        scored_network,\n",
    "        n_per_gene=n_references,\n",
    "        relation_filter=(\"reference\", \"citation\"),\n",
    "    )\n",
    "    top_references = enrich_with_epmc_fulltext(\n",
    "        top_references,\n",
    "        delay=0.1,\n",
    "        include_xml=True,\n",
    "    )\n",
    "    top_references[\"score\"] = top_references[\"composite_score\"]\n",
    "    preview = top_references.assign(\n",
    "        abstract_chars=top_references[\"abstract_text\"].fillna(\"\").str.len(),\n",
    "        plain_text_chars=top_references[\"plain_text\"].fillna(\"\").str.len(),\n",
    "    )\n",
    "    display(preview[[\n",
    "        \"gene_symbol\",\n",
    "        \"relation_primary\",\n",
    "        \"title\",\n",
    "        \"year\",\n",
    "        \"score\",\n",
    "        \"abstract_chars\",\n",
    "        \"plain_text_chars\",\n",
    "    ]])\n",
    "\n",
    "    saved_files = save_dataframe_rows_as_json(\n",
    "        top_references,\n",
    "        \"../data/papers\",\n",
    "        id_column=\"PMID\",\n",
    "        filename_prefix=\"PMID\",\n",
    "        indent=4,\n",
    "        drop_missing=True,\n",
    "    )\n",
    "    print(f\"Saved {len(saved_files)} top-scoring JSON files to ../data/papers\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22787654",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now once I have build a huge "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
