{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a665627e",
   "metadata": {},
   "source": [
    "# Sequence to function\n",
    "\n",
    "## The problem\n",
    "The problem we are working on is called sequence to function. Ideally this means we have a sequence and we infer some kind of function\n",
    "\n",
    "### The sequence: \n",
    "To me this means a few different things. \n",
    "- Gene sequence mutations\n",
    "- Gene and Protein Orthologs\n",
    "- Post-translational modification\n",
    "\n",
    "### The function: \n",
    "This also could mean a few things. My mind initially went to physiology, but now realized function could be alot broader. For example simple ligand enzyme binding could be related to funciton. The pathway that the protein is involved in, the regulation\n",
    "- A change in ligand binding interactions (Initial GPCR activation)\n",
    "- A change in metabolitic secondary activity (GPCRs downstream)\n",
    "\n",
    "## The solution\n",
    "When thinking about a solution to the hackathon problem. Ryan has suggested, could we start with building a full profile with only two data categories (for the sequence). To me this means: \n",
    "- Gene name, mutation/ortholog\n",
    "- Protein name, post-translational modification\n",
    "\n",
    "Given those could we pull out somekind of knowledge graph that will allow us to relate it to currently known research. This is where we will need the __agent__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80e1f8c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.fetch_data import fetch_uniprot_data, split_colon_list\n",
    "from scripts.epmc_utils import fetch_epmc, save_dataframe_rows_as_json\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "322e207b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists\n"
     ]
    }
   ],
   "source": [
    "#This simple script creates the JSON directory if it doesn't exist. We are adding this to .gitignore, maybe it is too heavy?\n",
    "if not os.path.isdir(\"../data/corpus/\"):\n",
    "    os.makedirs(\"../data/corpus/\", exist_ok = True)\n",
    "    print(\"Making a new directory\")\n",
    "else:\n",
    "    print(\"Directory already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d5d8fb90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenAge ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>entrez gene id</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>why</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GHR</td>\n",
       "      <td>growth hormone receptor</td>\n",
       "      <td>2690</td>\n",
       "      <td>GHR_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>GHRH</td>\n",
       "      <td>growth hormone releasing hormone</td>\n",
       "      <td>2691</td>\n",
       "      <td>SLIB_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>SHC1</td>\n",
       "      <td>SHC (Src homology 2 domain containing) transfo...</td>\n",
       "      <td>6464</td>\n",
       "      <td>SHC1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>POU1F1</td>\n",
       "      <td>POU class 1 homeobox 1</td>\n",
       "      <td>5449</td>\n",
       "      <td>PIT1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>PROP1</td>\n",
       "      <td>PROP paired-like homeobox 1</td>\n",
       "      <td>5626</td>\n",
       "      <td>PROP1_HUMAN</td>\n",
       "      <td>mammal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GenAge ID  symbol                                               name  \\\n",
       "0          1     GHR                            growth hormone receptor   \n",
       "1          2    GHRH                   growth hormone releasing hormone   \n",
       "2          3    SHC1  SHC (Src homology 2 domain containing) transfo...   \n",
       "3          4  POU1F1                             POU class 1 homeobox 1   \n",
       "4          5   PROP1                        PROP paired-like homeobox 1   \n",
       "\n",
       "   entrez gene id      uniprot     why  \n",
       "0            2690    GHR_HUMAN  mammal  \n",
       "1            2691   SLIB_HUMAN  mammal  \n",
       "2            6464   SHC1_HUMAN  mammal  \n",
       "3            5449   PIT1_HUMAN  mammal  \n",
       "4            5626  PROP1_HUMAN  mammal  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "307\n"
     ]
    }
   ],
   "source": [
    "genage_human = pd.read_csv(\"../data/raw/genage_human.csv\")\n",
    "display(genage_human.head())\n",
    "genes = genage_human['symbol'].values\n",
    "print(len(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6469f8c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching UniProt data for GHR\n",
      "Fetching UniProt data for GHRH\n",
      "Fetching UniProt data for SHC1\n",
      "Fetching UniProt data for POU1F1\n",
      "Fetching UniProt data for PROP1\n",
      "Fetching UniProt data for TP53\n",
      "Fetching UniProt data for TERC\n",
      "âœ— No UniProt result for TERC\n",
      "Fetching UniProt data for TERT\n",
      "Fetching UniProt data for ATM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pmids</th>\n",
       "      <th>dois</th>\n",
       "      <th>citation_titles</th>\n",
       "      <th>reviewed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GHR</td>\n",
       "      <td>P10912</td>\n",
       "      <td>Growth hormone receptor</td>\n",
       "      <td>MDLWQLLLTLALAGSSDAFSGSEATAAILSRAPWSLQSVNPGLKTN...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Growth hormone receptor and serum binding prot...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GHRH</td>\n",
       "      <td>P01286</td>\n",
       "      <td>Somatoliberin</td>\n",
       "      <td>MPLWVFFFVILTLSNSSHCSPPPPLTLRMRRYADAIFTNSYRKVLG...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Cloning and sequence analysis of cDNA for the ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SHC1</td>\n",
       "      <td>P29353</td>\n",
       "      <td>SHC-transforming protein 1</td>\n",
       "      <td>MDLLPPKPKYNPLRNESLSSLEEGASGSTPPEELPSPSASSLGPIL...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>A novel transforming protein (SHC) with an SH2...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>POU1F1</td>\n",
       "      <td>P28069</td>\n",
       "      <td>Pituitary-specific positive transcription fact...</td>\n",
       "      <td>MSCQAFTSADTFIPLNSDASATLPLIMHHSAAECLPVSNHATNVMS...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Cloning of the human cDNA for transcription fa...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>O75360</td>\n",
       "      <td>Homeobox protein prophet of Pit-1</td>\n",
       "      <td>MEAERRRQAEKPKKGRVGSNLLPERHPATGTPTTTVDSSAPPCRRL...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human Prop-1: cloning, mapping, genomic struct...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TP53</td>\n",
       "      <td>P04637</td>\n",
       "      <td>Cellular tumor antigen p53</td>\n",
       "      <td>MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLS...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Human p53 cellular tumor antigen: cDNA sequenc...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TERT</td>\n",
       "      <td>O14746</td>\n",
       "      <td>Telomerase reverse transcriptase</td>\n",
       "      <td>MPRAPRCRAVRSLLRSHYREVLPLATFVRRLGPQGWRLVQRGDPAA...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hEST2, the putative human telomerase catalytic...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ATM</td>\n",
       "      <td>Q13315</td>\n",
       "      <td>Serine-protein kinase ATM</td>\n",
       "      <td>MSLVLNDLLICCRQLEHDRATERKKEVEKFKRLIRDPETIKHLDRH...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>The complete sequence of the coding region of ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol uniprot_id                                       protein_name  \\\n",
       "0         GHR     P10912                            Growth hormone receptor   \n",
       "1        GHRH     P01286                                      Somatoliberin   \n",
       "2        SHC1     P29353                         SHC-transforming protein 1   \n",
       "3      POU1F1     P28069  Pituitary-specific positive transcription fact...   \n",
       "4       PROP1     O75360                  Homeobox protein prophet of Pit-1   \n",
       "5        TP53     P04637                         Cellular tumor antigen p53   \n",
       "6        TERT     O14746                   Telomerase reverse transcriptase   \n",
       "7         ATM     Q13315                          Serine-protein kinase ATM   \n",
       "\n",
       "                                            sequence pmids  dois  \\\n",
       "0  MDLWQLLLTLALAGSSDAFSGSEATAAILSRAPWSLQSVNPGLKTN...  None  None   \n",
       "1  MPLWVFFFVILTLSNSSHCSPPPPLTLRMRRYADAIFTNSYRKVLG...  None  None   \n",
       "2  MDLLPPKPKYNPLRNESLSSLEEGASGSTPPEELPSPSASSLGPIL...  None  None   \n",
       "3  MSCQAFTSADTFIPLNSDASATLPLIMHHSAAECLPVSNHATNVMS...  None  None   \n",
       "4  MEAERRRQAEKPKKGRVGSNLLPERHPATGTPTTTVDSSAPPCRRL...  None  None   \n",
       "5  MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLS...  None  None   \n",
       "6  MPRAPRCRAVRSLLRSHYREVLPLATFVRRLGPQGWRLVQRGDPAA...  None  None   \n",
       "7  MSLVLNDLLICCRQLEHDRATERKKEVEKFKRLIRDPETIKHLDRH...  None  None   \n",
       "\n",
       "                                     citation_titles  reviewed  \n",
       "0  Growth hormone receptor and serum binding prot...     False  \n",
       "1  Cloning and sequence analysis of cDNA for the ...     False  \n",
       "2  A novel transforming protein (SHC) with an SH2...     False  \n",
       "3  Cloning of the human cDNA for transcription fa...     False  \n",
       "4  Human Prop-1: cloning, mapping, genomic struct...     False  \n",
       "5  Human p53 cellular tumor antigen: cDNA sequenc...     False  \n",
       "6  hEST2, the putative human telomerase catalytic...     False  \n",
       "7  The complete sequence of the coding region of ...     False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Maybe here can go the parsing terms\n",
    "uniprot_data = fetch_uniprot_data(genes[0:9])\n",
    "display(uniprot_data)\n",
    "citation_list = split_colon_list(uniprot_data.citation_titles[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c25d4eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_citation_records = [\n",
    "    fetch_epmc(citation)\n",
    "    for citation in citation_list\n",
    "]\n",
    "uniprot_citations_df = pd.DataFrame(uniprot_citation_records)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bced0ed7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing citation record for  Growth hormone receptor and serum binding protein: purification, cloning and expression.\n",
      "Fetching EPMC metadata for: Growth hormone receptor and serum binding protein: purification, cloning and expression.\n",
      "Metadata fetched: {'PMC': '1718137', 'DOI': '10.1136/adc.81.5.378', 'PMID': '10519707', 'PMCID': 'PMC1718137', 'title': 'Growth hormone insensitivity: a widening diagnosis.', 'journal': None, 'year': 1999, 'source_url': 'https://europepmc.org/article/PMC/PMC1718137', 'source': 'MED'}\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'empty'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 6\u001b[39m\n\u001b[32m      3\u001b[39m uniprot_citation_record = fetch_epmc(citation)\n\u001b[32m      5\u001b[39m \u001b[38;5;66;03m# uniprot_citations_df = pd.DataFrame(uniprot_citation_record)\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m6\u001b[39m saved_uniprot_citations = \u001b[43msave_dataframe_rows_as_json\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m    \u001b[49m\u001b[43muniprot_citation_record\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m      8\u001b[39m \u001b[43m    \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m../data/corpus/\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      9\u001b[39m \u001b[43m    \u001b[49m\u001b[43mid_column\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mPMID\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     10\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfilename_prefix\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mEPMC_\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     11\u001b[39m \u001b[43m    \u001b[49m\u001b[43mindent\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m4\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     12\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdrop_missing\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m     13\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     14\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSaved \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(saved_uniprot_citations)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m UniProt citation JSON files.\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\mtarc\\Dev\\protein-seq-to-func\\notebooks\\..\\scripts\\epmc_utils.py:700\u001b[39m, in \u001b[36msave_dataframe_rows_as_json\u001b[39m\u001b[34m(df, directory, id_column, filename_prefix, indent, drop_missing)\u001b[39m\n\u001b[32m    677\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34msave_dataframe_rows_as_json\u001b[39m(\n\u001b[32m    678\u001b[39m     df: pd.DataFrame,\n\u001b[32m    679\u001b[39m     directory: \u001b[38;5;28mstr\u001b[39m | Path,\n\u001b[32m   (...)\u001b[39m\u001b[32m    684\u001b[39m     drop_missing: \u001b[38;5;28mbool\u001b[39m = \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[32m    685\u001b[39m ) -> \u001b[38;5;28mlist\u001b[39m[Path]:\n\u001b[32m    686\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    687\u001b[39m \u001b[33;03m    Persist each row of a DataFrame as an individual JSON file.\u001b[39;00m\n\u001b[32m    688\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m    698\u001b[39m \u001b[33;03m        list[pathlib.Path]: Paths to the files that were written.\u001b[39;00m\n\u001b[32m    699\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m700\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m df \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m \u001b[43mdf\u001b[49m\u001b[43m.\u001b[49m\u001b[43mempty\u001b[49m:\n\u001b[32m    701\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m []\n\u001b[32m    703\u001b[39m     target_dir = Path(directory)\n",
      "\u001b[31mAttributeError\u001b[39m: 'dict' object has no attribute 'empty'"
     ]
    }
   ],
   "source": [
    "for citation in citation_list:\n",
    "    print(\"Processing citation record for \",citation)\n",
    "    uniprot_citation_record = fetch_epmc(citation)\n",
    "\n",
    "    # uniprot_citations_df = pd.DataFrame(uniprot_citation_record)\n",
    "    saved_uniprot_citations = save_dataframe_rows_as_json(\n",
    "        uniprot_citation_record,\n",
    "        \"../data/corpus/\",\n",
    "        id_column=\"PMID\",\n",
    "        filename_prefix=\"EPMC_\",\n",
    "        indent=4,\n",
    "        drop_missing=True,\n",
    "    )\n",
    "    print(f\"Saved {len(saved_uniprot_citations)} UniProt citation JSON files.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
